### ISSUE TITLE
[DirectoryFunctions::CreateDirectoriesRecursively] The last directory is not created if there is no '/' at the end of the path

### ISSUE THREAD

**Issue by kawaii-flesh** [J-D-K/FsLib#2]
Hi! In the current implementation of `CreateDirectoriesRecursively` (there is no such issue with `CreateDirectory`), if the directory path doesn't end with a '/', the last directory in the path won't be created.
I only tested it in the Switch implementation, but as I understand, this also applies to the 3DS

For example:
- `"sdmc:/root/child"` – `child` won't be created.
- `"sdmc:/root/child/"` – `child` will be created.

Do you have any conventions regarding this? If not, could you please add something like this, or something more elegant?
```cpp
bool FsLib::CreateDirectoriesRecursively(const FsLib::Path &DirectoryPath)
{
    FsLib::Path FullDirectoryPath = DirectoryPath;
    size_t FullDirectoryPathLength = FullDirectoryPath.GetLength();
    if (FullDirectoryPathLength > 0 && FullDirectoryPath.CString()[FullDirectoryPathLength - 1] != '/')
        FullDirectoryPath += "/";

    // We don't really need to set any error strings here, because CreateDirectory will do that.
    size_t SlashPosition = FullDirectoryPath.FindFirstOf('/') + 1;
    while ((SlashPosition = FullDirectoryPath.FindFirstOf('/', SlashPosition)) != FullDirectoryPath.npos)
    {
        FsLib::Path CurrentDirectory = FullDirectoryPath.SubPath(SlashPosition);
        if (!FsLib::DirectoryExists(CurrentDirectory) && !FsLib::CreateDirectory(CurrentDirectory))
        {
            return false;
        }
        ++SlashPosition;
    }
    return true;
}
```

I love what you're doing!) Thank you!

**Comment by J-D-K** [comment]
I'll see what I can think of. Maybe it'll just be both in the end. I was considering making the Path closer to std::filesystem::path, which I'm pretty sure doesn't use trailing slashes. Adding operator/ too. Just gotta plan it out. It shouldn't be a lot of work because it's basically all there. It's just that I'm being pulled in five different directions right now.

**Comment by J-D-K** [comment]
Oh, and thanks. I'm basically breaking JKSM and JKSV down into pieces to make stuff easier to use and maintain. We'll see how and where it goes. Normally I wouldn't try making it easy to make certain _things_, but I already started fiddling with Gamecard stuff...

**Comment by J-D-K** [comment]
The work is already done. The changes I've made are that FsLib::Path will trim off trailing slashes from any path. operator+= and operator+ are both now operator/= and operator/. I'm still testing and making sure everything works right. It seems like it does, but I'm really thinking this part over:
```cpp 
// This is to trim trailing slashes from the end of the path.
size_t GetPathLength(const char *P)
{
    // This is to trim trailing slashes from the path.
    char *PathBegin = std::strchr(P, '/');
    if (!PathBegin)
    {
        // This might be bad...
        return std::strlen(P);
    }

    // Skip over the first slash so it can't be deleted by the loop.
    ++PathBegin;
    size_t PathLength = std::strlen(PathBegin);
    while (PathLength > 0 && PathBegin[PathLength - 1] == '/')
    {
        --PathLength;
    }

    return (PathBegin - P) + PathLength;
}
```
I'm really considering whether or not just returning strlen in the event a slash isn't found is a good idea. If FsLib isn't used correctly and someone tries to do something like:
```cpp
FsLib::Path JKSVPath = "sdmc:/JKSV/";
JKSVPath /= "/Pokemon Shield/";
```
The first path will result in `sdmc:/JKSV`, but the second will result in `sdmc:/JKSV//Pokemon Shield`. At the same time, I've got to consider where the training wheels get taken off so to speak. How much can I realistically take care of and handle before it's user error? I also changed the structure of CreateDirectoriesRecursively so the condition is checked at end which should take care of what you were talking about. I'll leave the issue open until it's tested better and pushed and I can truly consider it resolved.

**Comment by kawaii-flesh** [comment]
`sdmc:/JKSV//Pokemon Shield`
I think such a case should not be fixed "automatically" and should be covered in `Path::isValid`

**Comment by J-D-K** [comment]
Path has been updated to use operator/ and operator/=. I decided to correct for beginning and trailing slashes. I'm going to run some more quick tests to be sure this is working 100% as intended though. Using it for JKSV should be more than enough of a test after that.

As for checking paths for doubled up slashes and that kind of stuff, yeah, I can see it being useful, but I think at some point it becomes a programmer problem. FsLib will catch it in some form and will return false at some point and FsLib::GetErrorString() will have details. If I feed stdio or fstream a bad path, they will more or less do the same. I'm leaving this issue open until I'm 100% certain the new code works entirely as intended though.
